092-self-healing-using-cruise-control.md

```diff
@@ -0,0 +1,284 @@
+```
+
+The anomaly status would change based on how the healing progresses. The anomaly can transition to these possible status:
+* DETECTED - The anomaly is just detected and no action is yet taken on it 
+* IGNORED - Self healing is either disabled or the anomaly is unfixable
+* FIX_STARTED - The anomaly has started getting fixed
+* FIX_FAILED_TO_START - The proposal generation for the anomaly fix failed 
+* CHECK_WITH_DELAY - The anomaly fix is delayed 
+* LOAD_MONITOR_NOT_READY - Monitor which is monitoring the workload of a Kafka cluster is in loading or bootstrap state
+* COMPLETENESS_NOT_READY - Completeness not met for some goals. The completeness describes the confidence level of the data in the metric sample aggregator. If this is low then some Goal classes may refuse to produce proposals.
+
+We can also poll the `state` endpoint, with `substate` parameter as `executor`, to get details on the tasks that are currently being performed by Cruise Control.
+For example:
+```shell
+ExecutorState: {state: INTER_BROKER_REPLICA_MOVEMENT_TASK_IN_PROGRESS,
+ pending(20)/in-progress(0)/aborting(0)/finished(0)/total(20) inter-broker partition movements, 
+ completed(0)/total(0) bytes in MBs: 0.00%, max/min/avg concurrent inter-broker partition movements per-broker: 5/5/5.00, 
+ triggeredSelfHealingTaskId: 4b1ed191-50d7-4aa1-806b-b8c5c0d679fe, triggeredTaskReason:
+ Self healing for DISK_FAILURE: {
+	Disk /var/lib/kafka/data-1/kafka-log2 on broker 2 failed at 2024-12-09T09:49:46Z
+ }}
+```
+
+### Enabling self-healing configuration in the Strimzi operator
+
+Every configuration related to self-healing is currently disabled in the operator, and we will need to enable users altering some of those properties.
+The setting of these properties will be placed behind a feature gate.
+
+### Feature Gate
+
+The new feature gate will be called `UseSelfHealing`.
+It will be introduced in an alpha state and will be disabled by default.
+
+We are using a feature gate for this feature due to various reasons:
+* We want to allow users to try this feature and test it.
+* We would like to know more about the positive/negatives of this feature by testing it in real deployments.
+* Getting feedback from the community would allow us to understand how we can improve this even more.
+
+### Enabling self-healing in the Kafka custom resource
+
+When the feature gate is enabled, the operator will set the `self.healing.enabled` Cruise Control configuration to `true` and allow the users to set other appropriate `self-healing` related [configurations](https://github.com/linkedin/cruise-control/wiki/Configurations#selfhealingnotifier-configurations).
+
+Here is an example of how the configured Kafka custom resource could look:
+```yaml
+apiVersion: kafka.strimzi.io/v1beta2
+kind: Kafka
+metadata:
+  name: my-cluster
+spec:
+  kafka:
+    # ...
+  cruiseControl:
+    config:
+      # configuration to enable self-healing for all anomalies
+      self.healing.enabled: true
+```
```

Comment:
These two things seem to contradict each other ...
* Does the `UseSelfHealing` FG enable the self-healing as sugested in the first paragraph?
* Or does it allow the user to enable self-healing as suggested by the example YAML?

Reply:
I would say the first one. FG enables the self healing. I will remove this

Reply:
That would mean you would force the self-healing one day on all users even if they don't want it. That does not seem to be the right way, or?

Reply:
I also find FG and this configuration a bit confusing. If FG is not enabled, would this configuration be allowed by user? What would happen, if user sets this configuration without enabling FG?

Reply:
My take on this is that the user should enable self healing (and all the corresponding properties/configuration) which would be anyway ignored if the FG is not enabled. This way we are putting the "enablement and configuration of self-healing" behind a FG.